### PR TITLE
feat: concat indexes in the same object

### PR DESCRIPTION
**Pull request by jeqo**
By having indexes together, the number of objects stored in remote tier got reduced from 4 or 5 to 1. Range fetch used to get specific indexes.

### CODE REVIEW

File: core/src/test/java/io/aiven/kafka/tieredstorage/manifest/SegmentManifestV1SerdeTest.java
Comment (ivanyu) [inline core/src/test/java/io/aiven/kafka/tieredstorage/manifest/SegmentManifestV1SerdeTest.java:1]: It would be great to have the test for serialization/deserialization without the transaction index. Theoretically, it should serialize as `null` and deserialize just OK, but it would be great to check.

File: core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+import io.aiven.kafka.tieredstorage.storage.BytesRange;
+
+import static org.apache.kafka.server.log.remote.storage.RemoteStorageManager.IndexType;
+
+public class SegmentIndexesV1Builder {
+    final Map<IndexType, SegmentIndexV1> indexes = new HashMap<>(IndexType.values().length);
+    int currentPosition = 0;
+
+    public SegmentIndexesV1Builder add(final IndexType indexType, final BytesRange range) {
```
Comment (ivanyu) [inline core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java:30]: It seems from the range we only need the size. Maybe we should pass it instead? This will also eliminate the problem of the need to check for overlapping ranges, extra tests for this, etc.

File: core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java [hunk 2]
```diff
@@ -301,35 +355,30 @@ private void uploadSegmentLog(final RemoteLogSegmentMetadata remoteLogSegmentMet
...
-                                 final SegmentEncryptionMetadata encryptionMetadata,
-                                 final SegmentCustomMetadataBuilder customMetadataBuilder)
-        throws StorageBackendException, IOException {
-        TransformChunkEnumeration transformEnum = new BaseTransformChunkEnumeration(index);
+    private InputStream transformIndex(final IndexType indexType,
+                                       final InputStream index,
+                                       final int size,
+                                       final SegmentEncryptionMetadata encryptionMetadata,
+                                       final SegmentIndexesV1Builder segmentIndexBuilder) {
+        TransformChunkEnumeration transformEnum = new BaseTransformChunkEnumeration(index, Integer.MAX_VALUE / 2);
```
Comment (ivanyu) [inline core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java:363]: Will this work?
```suggestion
        TransformChunkEnumeration transformEnum = new BaseTransformChunkEnumeration(index, size);
```

File: core/src/test/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/ChunkCacheTest.java
Comment (ivanyu) [inline core/src/test/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/ChunkCacheTest.java:1]: [Nit] Would be great to not have these extra reformattings. They seem valid (?) but should probably be a separate PR.

File: core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java [hunk 3]
```diff
@@ -415,6 +469,8 @@ public InputStream fetchIndex(final RemoteLogSegmentMetadata remoteLogSegmentMet
             }
             final DetransformFinisher detransformFinisher = new DetransformFinisher(detransformEnum);
             return detransformFinisher.toInputStream();
+        } catch (final RemoteResourceNotFoundException e) {
+            throw e;
```
Comment (ivanyu) [inline core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java:473]: This is redundant

File: core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java [hunk 4]
```diff
@@ -401,8 +450,13 @@ public InputStream fetchIndex(final RemoteLogSegmentMetadata remoteLogSegmentMet
...
             final var segmentManifest = fetchSegmentManifest(remoteLogSegmentMetadata);
 
-            final var key = objectKey(remoteLogSegmentMetadata, ObjectKeyFactory.Suffix.fromIndexType(indexType));
-            final var in = fetcher.fetch(key);
+            final var key = objectKey(remoteLogSegmentMetadata, ObjectKeyFactory.Suffix.INDEXES);
+            final var segmentIndex = segmentManifest.segmentIndexes().segmentIndex(indexType);
+            if (segmentIndex == null) {
+                throw new RemoteResourceNotFoundException("Index " + indexType + " not found on " + key);
+            }
+            final var range = BytesRange.ofFromPositionAndSize(segmentIndex.position(), segmentIndex.size());
```
Comment (ivanyu) [inline core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java:458]: We can probably add `segmentIndex.range()` or something of this sort, but your call :)

File: core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java (hunk 3 above)
Comment (jeqo) [reply core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java:473]: not really. As the throw is within a try-catch, is captured by the last `catch Exception` and turned into a RSMException -- breaking the exception handling on the broker side.

File: core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java (hunk 2 above)
Comment (jeqo) [reply core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java:363]: Actually, it does! Thanks for the suggestion.

File: core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java (hunk 4 above)
Comment (jeqo) [reply core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java:458]: Sure, good idea. Will add it.

File: core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexV1.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ */
+
+package io.aiven.kafka.tieredstorage.manifest;
+
+import java.util.Objects;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class SegmentIndexV1 implements SegmentIndex {
```
[the remark this answers is not in the document]
Comment (jeqo) [reply core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexV1.java:26]: Added #417

File: core/src/test/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/ChunkCacheTest.java
Comment (AnatolyPopov) [reply core/src/test/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/ChunkCacheTest.java:1]: I believe if we want to have it like this we should do it once and for all and hide the commit from the history, like we did in some other places.

File: core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        switch (indexType) {
+            case OFFSET:
+                if (!indexes.isEmpty() || currentPosition != 0) {
+                    throw new IllegalStateException("Indexes not empty, offset index must be first");
+                }
```
[the remark this answers is not in the document]
Comment (jeqo) [reply core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java]: Yeah, wasn't sure order if order would be relevant here. Indeed, it may not matter. Will remove it.

File: core/src/test/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/ChunkCacheTest.java
Comment (jeqo) [reply core/src/test/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/ChunkCacheTest.java:1]: Sure, leaving this out. We can follow @AnatolyPopov approach on another PR.

File: core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java (hunk 3 above)
Comment (ivanyu) [reply core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java:472]: Yes, you're right!

File: core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ivanyu) [reply core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java]: I think the whole switch could be replaced with
```java
        indexes.put(indexType, new SegmentIndexV1(currentPosition, size));
        currentPosition += size;
```

File: core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+package io.aiven.kafka.tieredstorage.manifest;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.kafka.server.log.remote.storage.RemoteStorageManager.IndexType;
+
+public class SegmentIndexesV1Builder {
+    final Map<IndexType, SegmentIndexV1> indexes = new HashMap<>(IndexType.values().length);
+    int currentPosition = 0;
```
Comment (ivanyu) [inline core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java:26]: These guys should be private, I guess

File: core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                indexes.put(IndexType.LEADER_EPOCH, new SegmentIndexV1(currentPosition, size));
+                currentPosition += size;
+                break;
+            case TRANSACTION:
+                indexes.put(IndexType.TRANSACTION, new SegmentIndexV1(currentPosition, size));
+                currentPosition += size;
+                break;
+            default:
+                throw new IllegalStateException("Unknown index type");
+        }
```
Comment (ivanyu) [inline core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java:55]: I think the whole switch could be replaced with
```suggestion
        indexes.put(indexType, new SegmentIndexV1(currentPosition, size));
        currentPosition += size;
```

Comment (jeqo) [reply core/src/main/java/io/aiven/kafka/tieredstorage/manifest/SegmentIndexesV1Builder.java:55]: of course, 🤦🏽

**Review by ivanyu — CHANGES_REQUESTED** [review]
LGTM, @jeqo feel free to squash

**Review by ivanyu — APPROVED** [review]
